File: 189. Rotate Array/189. Rotate Array.cpp
/*
 * Problem: 189. Rotate Array
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/rotate-array/
 * Language: cpp
 * Date: 2026-04-26
 */

class Solution {
public:
    void rotate(vector<int>& nums, int k) {

        int n = nums.size();

        // Create vector with size n because we already know
        // exact positions where elements must be placed.
        // If it were empty + push_back(), elements would only
        // be added at end, not directly inserted at index.
        vector<int> demo(n);

        for (int i = 0; i < n; i++) {

            // New index after right rotation
            int index = (i + k) % n;

            // Direct assignment needs valid pre-sized indexes
            demo[index] = nums[i];
        }

        // Replace original array with rotated array
        nums = demo;
    }
};

File: 189. Rotate Array/189. Rotate Array.py
#
# Problem: 189. Rotate Array
# Difficulty: Medium
# Link: https://leetcode.com/problems/rotate-array/
# Language: python
# Date: 2026-04-26


class Solution(object):
    def rotate(self, nums, k):

        n = len(nums)

        # We create list with size n because elements will be placed
        # directly at specific indexes.
        # Empty list works with append(), but here we need demo[index].
        # So indexes must already exist.
        demo = [0] * n

        for i in range(n):

            # Calculate rotated position for current element
            index = (i + k) % n

            demo[index] = nums[i]

        # nums = demo would only change local reference.
        # nums[:] = demo replaces contents of original list,
        # so changes reflect outside the function too.
        nums[:] = demo
